modules/collection/programs/fish.nix

```
@@ -0,0 +1,98 @@
+{
+  pkgs,
+  lib,
+  config,
+  ...
+}: let
+  inherit (lib.options) mkOption mkEnableOption mkPackageOption;
+  inherit (lib.strings) typeOf concatMapAttrsStringSep concatMapStringsSep splitString;
+  inherit (lib.modules) mkIf;
+  inherit (lib.types) either str path oneOf attrsOf nullOr bool;
+  inherit (lib.attrsets) mapAttrs' nameValuePair isDerivation;
+
+  cfg = config.rum.programs.fish;
+  env = config.environment.sessionVariables;
+
+  toFish = strOrPath: fileName:
+    if (typeOf strOrPath) == "string"
+    then pkgs.writers.writeFish fileName strOrPath
+    else if isDerivation strOrPath
+    then strOrPath
+    else throw "Input is of invalid type ${typeOf strOrPath}, expected `path` or `string`.";
+
+  toFishFunc = strOrPath: funcName:
+    toFish (
+      if (typeOf strOrPath) == "string"
+      then ''
+        function ${funcName};
+        ${concatMapStringsSep "\n" (line: "\t${line}") (splitString "\n" strOrPath)}
+        end
+      ''
+      else strOrPath
+    ) "${funcName}.fish";
+in {
+  options.rum.programs.fish = {
+    enable = mkEnableOption "fish";
+
+    package = mkPackageOption pkgs "fish" {};
+
+    config = mkOption {
+      default = null;
+      type = nullOr (either str path);
+      description = ''
+        The main configuration for fish, written verbatim to `.config/fish/config.fish`.
+      '';
+    };
+
+    exportEnvVars = mkOption {
+      default = true;
+      type = bool;
+      description = ''
+        Wether to export environment variables set in ''${config.environment.systemVariables}.
+      '';
+    };
+
+    prompt = mkOption {
+      default = null;
+      type = nullOr (either str path);
+      description = ''
+        Your fish prompt, written to `.config/fish/functions/fish_prompt.fish`.
+        It follows the behaviour of `rum.programs.fish.functions`.
+      '';
+    };
+
+    functions = mkOption {
+      default = {};
+      type = attrsOf (oneOf [str path]);
+      description = ''
+        A fish function which is being written to `.config/fish/functions/<name>.fish`.
+
+        If the input value is a string, its contents will be wrapped
+        inside of a function declaration, like so:
+        ```fish
+            function <name>;
+                <function body>
+            end
+        ```
+
+        Otherwise you are expected to handle that yourself.
+      '';
+    };
+  };
```

```suggestion
    functions = mkOption {
      default = {};
      type = attrsOf (oneOf [str path]);
      description = ''
        A fish function which is being written to `.config/fish/functions/<name>.fish`.

        If the input value is a string, its contents will be wrapped
        inside of a function declaration, like so:
        ```fish
            function <name>;
                <function body>
            end
        ```

        Otherwise you are expected to handle that yourself.
      '';
      example = literalExample ''
        function fish_prompt -d "Write out the prompt"
              # This shows up as USER@HOST /home/user/ >, with the directory colored
              # $USER and $hostname are set by fish, so you can just use them
              # instead of using `whoami` and `hostname`
              printf '%s@%s %s%s%s > ' $USER $hostname \
                  (set_color $fish_color_cwd) (prompt_pwd) (set_color normal)
        end
      '';
    };
  };
```

See above (example taken from https://fishshell.com/docs/current/cmds/fish_prompt.html#example).